### DISCUSSION TITLE
[Roadmap] Astro v1.0

### DISCUSSION THREAD

**Discussion by FredKSchott** [withastro/roadmap#1] [Proposal]
# Summary

Astro v1.0 is our next goal, which means cleaning up and finalizing any APIs for a stable release. This document serves as a place to collect and organize all v1.0 proposals into a holistic view, so that the final API can be viewed in total. Larger tasks will be discussed and finalized in their own RFCs (linked to below).

# v1.0 API Overview

## ✅ XSS Protection

<details>
  <summary>Expand</summary>

### Automatic Escaping 💪

```
<div foo="aaa">bbb</div>          ->  <div foo="aaa">bbb</div>
<div foo={'"aaa\" />'}>bbb</div>  ->  <div foo="&quot;aaa&quot; />">bbb</div>
<div>{'</div>'}</div>             ->  <div>&lt;/div&gt;</div>
<div>{'<script>alert("ohno")</script>'}</div>  ->  <div>&lt;script&gt;alert("ohno")&lt;/script&gt;</div>
{someHtml}          ->  &lt;script&gt;alert("ohno")&lt;/script&gt;
<SomeVariable />    ->  <somevariable></somevariable>
<{SomeVariable} />  ->  Syntax Error
```

- [X] Owner: @natemoo-re 
- [X] Discussion: https://github.com/withastro/astro/issues/1828 (needs to be migrated to RFC or discussion in new repo)
- [x] RFC: TODO (everyone bought into the concept, mainly to flesh out implementation. could be prototyped first and then RFC just confirms prototype behavior before PR)
- [x] Implementation: TODO

### Explicit Unescaped HTML Fallback  💪

```
<div set:html={unescapedHtml} />
<div set:text={someText} />
```

- [X] Owner: @natemoo-re 
- [X] Discussion: https://github.com/withastro/astro/issues/1827 (needs to be migrated to RFC or discussion in new repo)
- [X] RFC: Link needed
- [x] Implementation: TODO

</details>

## ✅  SSR Required Changes

<details>
  <summary>Expand</summary>

### Enforce static of client directives  💪

- [X] Owner: @matthewp 
- [x] Discussion: https://github.com/withastro/rfcs/discussions/30 
- [x] RFC: Link needed
- [x] Implementation: Link needed, Behind flag

### Remove Astro.resolve 💪

- [X] Owner: @matthewp 
- [x] Discussion: https://github.com/withastro/rfcs/discussions/32
- [x] RFC: Link needed
- [x] Implementation: TODO

### Move Hydration scripts back to being inline 💪💪

- [X] Owner: @matthewp 
- [x] Discussion: https://github.com/withastro/rfcs/discussions/36
- [X] RFC: Not needed, implementation detail
- [x] Implementation: Link needed, Behind flag

### Remove HTML post-processing 💪💪

- improve build performance, align static site generation to match SSR requirements.
- [X] Owner: @matthewp 
- [X] Discussion: Not needed, implementation detail
- [X] RFC: Not needed, implementation detail
- [x] Implementation: Link needed, Behind flag, needs polish

### Where do assets like images live? `public/` or `src/`?

Needs reframing? This was tackled by other proposals for the SSR build problem, but we may need other improvements for things like icons, etc.

- [X] Owner: @FredKSchott  
- [x] Discussion: TODO
- [x] RFC: https://github.com/withastro/rfcs/pull/106
- [x] Implementation: N/A

### Resolve Style Loading

- [X] Owner: @matthewp 
- [x] Discussion: https://github.com/withastro/rfcs/discussions/28
- [x] RFC: https://github.com/withastro/rfcs/blob/main/active-rfcs/0000-style-unification.md
- [x] Implementation: Done, behind experimental static build flag

### Audit Build Performance for Large Sites 💪

- [x] Achieve linear performance in builds (tracked in "SSR Required Changes")
- [x] Owner: @matthewp 
- [x] Benchmark: https://github.com/seancdavis/ssg-build-performance-tests

</details>

## ✅  Finalize `<head>`, `<body>`, and `<html>` Behavior 💪

<details>
  <summary>Expand</summary>

- [X] Owner: @natemoo-re  
- [x] RFC: https://github.com/withastro/rfcs/pull/26
- [x] Implementation: TODO

</details>

## ✅ Finalize Component Style and Script API 💪💪

<details>
  <summary>Expand</summary>

- [X] Owner: @natemoo-re
- [x] RFC: https://github.com/withastro/rfcs/pull/41
- [x] RFC: https://github.com/withastro/rfcs/pull/150
- [x] Implementation: https://github.com/withastro/astro/pull/2961

</details>

## ✅ Config Naming Audit 💪

<details>
  <summary>Expand</summary>

- Clean up confusing config, make sure all options are consistent and self-evident.
- [X] Owner: @natemoo-re 
- [X] Discussion: [Discord](https://discord.com/channels/830184174198718474/872579324446928896/954127837726646322)
- [X] RFC: https://github.com/withastro/rfcs/pull/154
- [X] Implementation: https://github.com/withastro/astro/pull/2962

</details>

## ✅ Finalize Default Renderer/Framework Behavior 💪💪

<details>
  <summary>Expand</summary>

To revisit and finalize two things: should every framework renderer come installed with Astro, and should every renderer be enabled by default?

- [X] Owner: @FredKSchott
- [X] RFC: #138 
- [X] Implementation: https://github.com/withastro/astro/pull/2820

</details>

## ✅ Finalize Default Sass/PostCSS/Tailwind Behavior 💪💪

<details>
  <summary>Expand</summary>

To revisit and finalize two things: should CSS addons come installed with Astro, and should every CSS addon be enabled by default?

- [x] Owner: @FredKSchott 
- [X] RFC: #138 
- [X] Implementation: https://github.com/withastro/astro/pull/2820

</details>

## ✅ Finalize `Astro.fetchContent()` API 💪

<details>
  <summary>Expand</summary>

Rendering template as a part of fetchContent causes infinite loop. Implementation was mostly solved by v0.21 `<Component />` export, but naming should be finalized. Remove old APIs and any properties that should no longer be used.

- [x] Owner: @FredKSchott 
- [x] Discussion: https://github.com/withastro/rfcs/discussions/5
- [x] RFC: https://github.com/withastro/rfcs/pull/147
- [x] Implementation: https://github.com/withastro/astro/pull/2862
</details>

## ✅ Finish `<Code>` prism->shiki migration

<details>
  <summary>Expand</summary>

- [x] Owner: @bholmesdev
- [x] RFC: [Accepted RFC](https://github.com/withastro/rfcs/blob/main/proposals/0004-replace-prism-with-shiki.md)
- [x] Implementation: Done

</details>

## ✅ `Astro.request` should be a real `Request`

<details>
  <summary>Expand</summary>

- [x] Owner: @matthewp 
- [x] RFC: [Accepted RFC](https://github.com/withastro/rfcs/blob/main/proposals/0018-astro-request.md)
- [x] Implementation: [PR](https://github.com/withastro/astro/pull/2913)

</details>

## v1.0.0 Beta => Stable

See [**[Roadmap] Astro v1.0.0 Beta => Stable**](https://github.com/withastro/rfcs/discussions/161)

**Comment by matthewp** [comment]
> wherever possible, removing situations where there are multiple ways to do the same thing (ex: <link> for style vs. import './foo.css' in frontmatter)

I don't think it would be technically feasible to remove either of those ways (not to speak of whether it is a good idea to do so). Do you mean to recommend one over the other?

**Reply by FredKSchott** [reply]
yup, the "wherever possible" is relevant there. That example may not be possible to resolve, but still worth a convo and I think that there are a few other "two ways to do the same thing" scenarios that we should also discuss and finalize.

**Comment by jonathantneal** [comment]
Re: Explicit Unescaped HTML Fallback

During our last discussion, I liked the idea of providing a built-in `<RawHTML>` component (or `<Fragment>` enhancement) for injecting HTML.

```astro
---
const content = "<h1>Hello world!</h1>";
---
<main><Fragment raw={content} /></main><!-- prints unescaped HTML -->
<main>{content}</main><!-- prints escaped HTML -->
```

Naming TBD.

**Reply by FredKSchott** [reply]
Please add to the relevant discussion thread! Keeping this discussion focused on overall orchestration and not any one individual item.

**Comment by jonathantneal** [comment]
Re: Directives Naming Audit

Minor nit: This should affect naming of the `@component` proposal.

**Comment by jonathantneal** [comment]
Re: `Astro.createComponent`

Opinion: We already have a way to create Astro components; `.astro` files.

I think there are 2 reasons one would reach for `Astro.createComponent`.

- Creating components in frontmatter. **Alternative**: Allow AstroHTML in Frontmatter again.
- Astro superpowers. Scoping. Hoisting. Etc. **Alternative**: Allow these features from other built-in utilities.

**Reply by FredKSchott** [reply]
We should chat about this one, I'd really like to ship v1.0 without it if we can and I can't help but think that Svelte and Vue seem to get by just fine without this feature. But I'd also love to hear more if I'm missing some valuable use-case.

**Reply by jonathantneal** [reply]
Sure thing. We probably agree, as I don’t think we need this feature. I felt enough community members missed it that it warranted a note.

The bit of my 'opinion' was that we already have a way to do this. The 'I think' bit was my interpretation of community feedback, and the 'alternative' bits were my suggestions to address that feedback.

My bias is that I don’t miss it. However, if AstroHTML in Frontmatter became a technical reality again, I’d support it. 🤷

**Comment by jonathantneal** [comment]
Re: Explicit Unescaped HTML Fallback

> (needs to be migrated to RFC or discussion in new repo)

It needs to be a new discussion. The related RFCs were not accepted.

**Disclaimer: I will now reference the duplicate notes from the related RFCs and my own flawed memory.**

---

**Nearly Settled**: Escaping by default

We came around on the need to escape strings and variables within expressions as the default. I believe this will pass easily now.

---

**Not Settled**: Escaping interface

We did not agree on the escape hatch API. I believe the suggested API methods for escape were an attribute, a utility, and an element.

---

**Not Settled**: Escaping effect

We did not agree on how escapes would be handled. I believe (at least?) one of us expected sanitization _was_ the escape, (at least?) two of us expected the API to support sanitization, and (at least?) one of us did not want any kind of sanitization at all.

---

**Nearly Settled**: Naming

There was consensus to avoid DOM property terminology in the API (i.e. `textContent`, `innerHTML`).

**Reply by FredKSchott** [reply]
> It needs to be a new discussion. The related RFCs were not accepted.

Ah no, I meant "migrated to RFC" as in a new, open RFC Pull Request. If @natemoo-re is able to migrate his old RFC to the new RFC format, then great. Otherwise, copying over as-is as a Discussion would be a fine short-term move just to keep the discussion going.

**Comment by Siilwyn** [comment]
What about HTML minification? Specifically stripping whitespace.

**Comment by jonathantneal** [comment]
I believe we are switching out `<slot>` for an alternative API. Could we add that here?

**Comment by Igloczek** [comment]
I'd love to see some clarification regarding the styles optimizations:
1. How to handle inlining? Usage of `set:html` might be an option, but at least for now, it doesn't work in all cases, for example ignore styles coming from Vue components.
2. How to handle async CSS?
3. How to merge stylesheets together?
3. How to have some control over the place where the styles `<link>` tags will be injected?

So basically all things required to use critical CSS with Astro, without some custom solutions running post-build.
